### DISCUSSION TITLE
Initializing clientConfig from a list of machines

### DISCUSSION THREAD

**Discussion by bsergean** [anacrolix/torrent#541] [Q&A]
Hi there,

I have a hard-coded list of string representing address:port (say host1:8080, host2:8082, etc...), and I'd like to use that to configure my client.

I see:
```
    DhtStartingNodes func(network string) dht.StartingNodesGetter
```

And by looking at the github repo for dht:

```
type StartingNodesGetter func() ([]Addr, error)
```

But I don't really know how to connect this. Do I need to set config.DhtStartingNodes to a function which return a list of Addr, and an error, and take as an input a string (network) ?

```
func MyFunction(network string) dht.StartingNodesGetter {
   return []Addr, nil // say if there was no error
}

// then
config.DhtStartingNodes = MyFunction
```

Edit: Addr seems to be defined as is:
```
// Used internally to refer to node network addresses. String() is called a
// lot, and so can be optimized. Network() is not exposed, so that the
// interface does not satisfy net.Addr, as the underlying type must be passed
// to any OS-level function that take net.Addr.
type Addr interface {
    Raw() net.Addr
    Port() int
    IP() net.IP
    String() string
    KRPC() krpc.NodeAddr
}
```

The unittest does this actually:

```
	clientConfig.DhtStartingNodes = func(string) dht.StartingNodesGetter { return func() ([]dht.Addr, error) { return nil, nil } }
```

Thanks !

**Comment by bsergean** [comment]
This now works for me !
We could probably cache the addr resolution or something similar.

```
	// DHT configuration
        hosts := "host1,host2,host3"
	clientConfig.NoDHT = false
	clientConfig.DhtStartingNodes = func(network string) dht.StartingNodesGetter {
		return func() ([]dht.Addr, error) {
			var addrs = make([]dht.Addr, 0)
			seeds := strings.Split(hosts, ",")
			for _, seed := range seeds {
				log.Printf("DHT: Resolving %s address %s", network, seed)
				if network == "udp" {
					var addr *net.UDPAddr
					addr, err := net.ResolveUDPAddr(network, seed + ":80")
					if err != nil {
						log.Errorf("Cannot resolve %s address %s", network, seed)
						continue
					}
					addrs = append(addrs, dht.NewAddr(addr))
				}
				if network == "tcp" {
					var addr *net.TCPAddr
					addr, err := net.ResolveTCPAddr(network, seed + ":80")
					if err != nil {
						log.Errorf("Cannot resolve %s address %s", network, seed)
						continue
					}
					addrs = append(addrs, dht.NewAddr(addr))
				}
			}
			return addrs, nil
		}
	}

```

**Reply by bsergean** [reply]
OK one problem is that I need a port number (:80) when passing in an address.

```
$ go run cmd/resolve.go www.google.com
2021/08/02 09:56:27 Cannot resolve address: www.google.com
$ go run cmd/resolve.go www.google.com:8080
2021/08/02 09:57:36 address: 216.58.212.100:8080
```

```
$ cat cmd/resolve.go 
package main

import "os"
import "log"
import "net"

func main() {
	addr, err := net.ResolveTCPAddr("tcp", os.Args[1])
	if err != nil {
		log.Printf("Cannot resolve address %s\n", os.Args[1])
	} else {
		log.Printf("address: %v\n", addr)
	}
}
```

**Reply by anacrolix** [reply]
You'll see that the default StartingNodesGetter, GlobalBootstrapAddrs, makes use of a DNS cache internal to the package. This is necessary because starting DHT traversals on servers that have no useful entries in the routing table will convoy on the starting nodes. Let me know if you have any suggestions about maybe exposing this, or perhaps you want to do your own caching for your own function.

On the port: You might do just IP DNS resolution without the port? I'm not sure what the dht repo should do about this. I don't think DHT addrs have any meaning without a port.

Thank you for sharing your interactions with the repo(s)!

**Reply by bsergean** [reply]
Thanks for making this library in the first place !

```
 dht server on [::]:6881 completed bootstrap (dht.TraversalStats{NumAddrsTried:60, NumResponses:0})
```

I've now set the port to the one used by the other seeds who I believe run a DHT server (with libtorrent) / but I still see NumResponses = 0 ... which makes me think they cannot be reached.

I'll look at how GlobalBootstrapAddrs works, I don't know if it could be started with a list of host names.

**Reply by anacrolix** [reply]
Just to confirm, you are trying to bootstrap your (implicit torrent.Client) DHT servers from these hosts? Otherwise you might just be looking to add those hosts as peers for the torrents, which is something else.

**Comment by bsergean** [comment]
Yes. Basically I am trying to emulate C++ code which add DHT nodes to a libtorrent session.

for (auto & seed : seeds) {
    session->add_dht_node( std::make_pair( seed.c_str(), port ) );
}

But you’re right that I could also just add those hosts as peers for the torrent which is pretty close. There are many instances of this program running, each one being a peer.

I’ve seen some APIs like ‘addPeer on a torrent, idk if they are external or internal yet.

**Reply by anacrolix** [reply]
I guess this is the equivalent then. You can also add the peers directly to the DHT server too: https://pkg.go.dev/github.com/anacrolix/dht#Server.AddNode.
